### PR TITLE
Feature/kmod tests

### PR DESCRIPTION
**Pull request by giladkl**

### CODE REVIEW

File: recorder/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import pytest
+import os
+import contextlib
+import psutil
+from subprocess import check_output, check_call
+from replayer.loader.system_call_loader import Loader
+
+KERNEL_MODULE = "record.ko"
+KERNEL_MODULE_PATH = f"../src/{KERNEL_MODULE}"
+READ_SIZE = 0xffff
```
Comment (maor10) [inline recorder/tests/conftest.py:11]: PEP8- Have two lines between top level functions

File: recorder/tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        
+        return Loader.from_buffer(content).load_system_calls()[:-2] # Last 2 syscalls are lseek and close 
+
+    return get
+        
+@pytest.fixture
+def record_syscalls_context():
+    @contextlib.contextmanager
+    def _record_syscalls():
+        with open("/proc/recorded_process", "w") as f:
```
Comment (maor10) [inline recorder/tests/conftest.py]: Style choice, but it's a bit weird to have the file open the entire time and needing to do flushes and seeks- why not open it before and after? Seems cleaner and more "direct" to what you want to do (imagine if you described this in pseudocode)

(this is also /much/ more common to see in python)

File: recorder/tests/stress/test_loading_unloading.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import pytest
+
+REPEAT_AMOUNT = 10
+
+@pytest.mark.parametrize("iteration", range(REPEAT_AMOUNT))
+def test_load_blast(kernel_module_context, whitelist_dmesg_context, iteration):
```
Comment (maor10) [inline recorder/tests/stress/test_loading_unloading.py]: I don't think parametrize is what you're looking for here; you're essentially testing the same thing n times. Each time would count as a separate test, which is not what you're shooting for here.
You should definitely have a for loop within the test causing the stressing.

File: recorder/tests/functionality/test_syscall.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import os
+import select
+
+def syscall_sum_copies(syscall):
+    return sum(len(copy.buffer) for copy in syscall.memory_copies)
+
+def test_read(kernel_module, record_syscalls_context, get_recorded_syscalls):
+    with open("/dev/urandom") as f:
+        with record_syscalls_context():
```
Comment (maor10) [inline recorder/tests/functionality/test_syscall.py]: Why are both `with`'s not in same line?

File: recorder/tests/functionality/test_syscall.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            ret = os.read(f.fileno(), 100)
```
Comment (maor10) [inline recorder/tests/functionality/test_syscall.py]: `ret` we are not in c my friend

File: recorder/tests/functionality/test_syscall.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import os
+import select
+
+def syscall_sum_copies(syscall):
```
Comment (maor10) [inline recorder/tests/functionality/test_syscall.py]: PEP8- Have two lines between top level functions

Comment (maor10) [inline recorder/tests/functionality/test_syscall.py]: Style Choice- try having your function names be in the  imperative- e.g. `get_sum_of_memory_copies_lengths`

File: recorder/tests/functionality/test_syscall.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    assert len(ret) == 100
+
+    assert len(syscalls) == 1
+    assert syscalls[0].name == "read"
+    assert syscalls[0].return_value == 100
+    assert syscall_sum_copies(syscalls[0]) == 100
+    
+def test_write(kernel_module, record_syscalls_context, get_recorded_syscalls):
+    
+    with open("/dev/null", "wb") as f:
```
Comment (maor10) [inline recorder/tests/functionality/test_syscall.py:23]: Same line for `with`s

File: recorder/tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+def kernel_module(kernel_module_context):
+    with kernel_module_context():
+        yield
+
+@pytest.fixture
+def whitelist_dmesg_context():
+    @contextlib.contextmanager
+    def _whitelist_dmesg_context(whitelist : list):
+        # Clear dmesg before yielding
+        check_output(["sudo", "dmesg", "-c"])
```
Comment (maor10) [inline recorder/tests/conftest.py]: namespaces are one honking good idea.
Not followed all the time, but specifically for subprocess it usually is :) 
Not critical

File: recorder/tests/conftest.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+@pytest.fixture
+def kernel_module(kernel_module_context):
+    with kernel_module_context():
+        yield
+
+@pytest.fixture
+def whitelist_dmesg_context():
+    @contextlib.contextmanager
+    def _whitelist_dmesg_context(whitelist : list):
```
Comment (maor10) [inline recorder/tests/conftest.py]: `whitelist: List[str]`

File: recorder/tests/conftest.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+            assert check_call(["sudo", "rmmod", KERNEL_MODULE]) == 0
+    return _kernel_module_context
+
+@pytest.fixture
+def kernel_module(kernel_module_context):
+    with kernel_module_context():
+        yield
+
+@pytest.fixture
+def whitelist_dmesg_context():
```
Comment (maor10) [inline recorder/tests/conftest.py:66]: why do you need a context manager? Why is this not just a yield fixture?

File: recorder/tests/conftest.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    def _whitelist_dmesg_context(whitelist : list):
+        # Clear dmesg before yielding
+        check_output(["sudo", "dmesg", "-c"])
+
+        yield
+
+        dmesg_after = check_output(["sudo", "dmesg", "-c"]).decode('ascii').strip()
+        for line in dmesg_after.split("\n"):
+            assert len([True for x in whitelist if x in line]) == 1, \
+                                        f"Line '{line}' not in whitelist"
```
Comment (maor10) [inline recorder/tests/conftest.py:77]: A few things:
1.  Are you trying to check whether any whitelist is /within/ a given line of output? Why would it be within and not exact? If it should be exact, the above lines could be switched with
```python
assert all([line in whitelist for line in dmesg_after.split("\n"])
```
2. `for x in` x is not a descriptive name, and you're looping through a list with meaning. Any other code within the loop is now less readable (this is in contrast to using `i` or `j` when looping through a `range`, which is common and expected)

File: recorder/tests/conftest.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+KERNEL_MODULE_PATH = f"../src/{KERNEL_MODULE}"
+READ_SIZE = 0xffff
+
+@pytest.fixture
+def get_recorded_syscalls():
+    def get():
+        fd = os.open("/proc/syscall_dumper", os.O_RDONLY | os.O_NONBLOCK)
+
+        content = b""
+        cur_read = os.read(fd, READ_SIZE)
```
Comment (maor10) [inline recorder/tests/conftest.py]: use full words when in da pyton

File: recorder/tests/conftest.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                f.seek(0)
+                f.write("0")
+                f.flush()
+    return _record_syscalls
+
+@pytest.fixture
+def kernel_module_context():
+    @contextlib.contextmanager
+    def _kernel_module_context():
+        assert check_call(["sudo", "insmod", KERNEL_MODULE_PATH]) == 0
```
Comment (maor10) [inline recorder/tests/conftest.py]: You don't need to assert, `check_call` raises an exception on non-zero exit statuses

**Review by maor10 — CHANGES_REQUESTED** [review]

File: recorder/tests/conftest.py (hunk 9 above)
Comment (giladkl) [reply recorder/tests/conftest.py]: This no work

File: recorder/tests/conftest.py (hunk 10 above)
Comment (giladkl) [reply recorder/tests/conftest.py:66]: How do i pass args to fixure of whitelist?

File: recorder/tests/conftest.py (hunk 11 above)
Comment (giladkl) [reply recorder/tests/conftest.py:77]: 1. Yea i need exact line -- because dmesg contains [<TIME>] Message
2. OK

Comment (giladkl) [reply recorder/tests/conftest.py:77]: 1. I NEED JUST PART OF LINE*

**Pull request merged**